core-wih/src/main/java/io/elimu/a2d2/corewih/OAuth2WorkItemHandler.java

```diff
@@ -0,0 +1,122 @@
+
+public class OAuth2WorkItemHandler implements WorkItemHandler {
+
+	private static final Logger LOG = LoggerFactory.getLogger(OAuth2WorkItemHandler.class);
+	
+	private final Pattern patToken = Pattern.compile(".*\"access_token\"\\s*:\\s*\"([^\"]+)\".*");
+	private final Pattern patExpires = Pattern.compile(".*\"expires_in\"\\s*:\\s*([^,]+).*");
+
+	@Override
+	public void executeWorkItem(WorkItem workItem, WorkItemManager manager) {
+		String clientId = (String) workItem.getParameter("clientId");
+		String scope = (String) workItem.getParameter("scope");
+		String username = (String) workItem.getParameter("username");
+		String password = (String) workItem.getParameter("password");
+		String grantType = (String) workItem.getParameter("grantType");
+		String tokenUrl = (String) workItem.getParameter("tokenUrl");
+		List<String> missingValues = new ArrayList<>();
+		if (clientId == null || "".equals(clientId.trim())) {
+			missingValues.add("clientId");
+		}
+		if (username == null || "".equals(username.trim())) {
+			missingValues.add("username");
+		}
+		if (password == null || "".equals(password.trim())) {
+			missingValues.add("password");
+		}
+		if (scope == null || "".equals(scope.trim())) {
+			scope = "offline_access";
+		}
+		if (grantType == null || "".equals(grantType.trim())) {
+			grantType = "password";
+		}
+		if (tokenUrl == null || "".equals(tokenUrl.trim())) {
+			missingValues.add("tokenUrl");
+		}
+		if (!missingValues.isEmpty()) {
+			throw new WorkItemHandlerException("Missing mandatory parameters: " + missingValues);
+		}
+		String content = "grant_type=" + grantType + "&client_id=" + clientId 
+				+ "&scope=" + scope + "&username=" + username 
+				+ "&password=" + password;
+	    BufferedReader reader = null; 
+	    HttpsURLConnection connection = null;
+	    String token = "";
+	    Map<String, Object> results = workItem.getResults();
+	    try {
+	        URL url = new URL(tokenUrl);
+	        connection = (HttpsURLConnection) url.openConnection();
+	        connection.setRequestMethod("POST");
+	        connection.setDoOutput(true);
+	        connection.setRequestProperty("Accept", "*/*");
+	        connection.setRequestProperty("Content-Type", "application/x-www-form-urlencoded");
+	        connection.setRequestProperty("Authorization", 
+	        		"Basic " + Base64.getEncoder().encodeToString((clientId + ":").getBytes()));
+	        PrintStream os = new PrintStream(connection.getOutputStream());
+	        os.print(content);
+	        os.close();
+	        reader = new BufferedReader(new InputStreamReader(connection.getInputStream()));
+	        String line = null;
+	        StringWriter out = new StringWriter(connection.getContentLength() > 0 ? connection.getContentLength() : 2048);
+	        while ((line = reader.readLine()) != null) {
+	            out.append(line);
+	        }
+	        String response = out.toString();
+	        Matcher matcher = patToken.matcher(response);
```

Comment:
Why are you using a pattern matcher instead of a Json parser here?

Reply:
Changed to use GSON. It seemed simpler when I was just extracting the access_token, but the more I added as an output the more it made sense to use a parser :)